### PR TITLE
*: turn on race detector for testing

### LINKED ISSUE #201: Race in distributor/protocols/tdp
Found in #198 by turning on the race detector.

```
$ go test -race github.com/coreos/torus/distributor/protocols/tdp
==================
WARNING: DATA RACE
Write at 0x00c42021a000 by goroutine 55:
  internal/race.WriteRange()
      /usr/local/go/src/internal/race/race.go:49 +0x42
  syscall.Read()
      /usr/local/go/src/syscall/syscall_unix.go:164 +0x9a
  net.(*netFD).Read()
      /usr/local/go/src/net/fd_unix.go:216 +0x146
  net.(*conn).Read()
      /usr/local/go/src/net/net.go:173 +0x96
  net.(*TCPConn).Read()
      <autogenerated>:69 +0x72
  github.com/coreos/torus/distributor/protocols/tdp.readConnIntoBuffer()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp.go:130 +0x95
  github.com/coreos/torus/distributor/protocols/tdp.(*Server).handlePutBlock()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp.go:175 +0x526
  github.com/coreos/torus/distributor/protocols/tdp.(*Server).handle()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp.go:108 +0x27a

Previous read at 0x00c42021a000 by goroutine 50:
  internal/race.ReadRange()
      /usr/local/go/src/internal/race/race.go:45 +0x42
  syscall.Write()
      /usr/local/go/src/syscall/syscall_unix.go:182 +0xaa
  net.(*netFD).Write()
      /usr/local/go/src/net/fd_unix.go:298 +0x187
  net.(*conn).Write()
      /usr/local/go/src/net/net.go:185 +0x96
  net.(*TCPConn).Write()
      <autogenerated>:70 +0x72
  github.com/coreos/torus/distributor/protocols/tdp.(*Conn).PutBlock()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/client.go:124 +0x9a3
  github.com/coreos/torus/distributor/protocols/tdp.TestPutBlock()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp_test.go:185 +0x27a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:610 +0xc9

Goroutine 55 (running) created at:
  github.com/coreos/torus/distributor/protocols/tdp.(*Server).serve()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp.go:80 +0x1c6

Goroutine 50 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:646 +0x52f
  testing.RunTests.func1()
      /usr/local/go/src/testing/testing.go:793 +0xb9
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:610 +0xc9
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:799 +0x4b5
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:743 +0x12f
  main.main()
      github.com/coreos/torus/distributor/protocols/tdp/_test/_testmain.go:80 +0x1b4
==================
--- FAIL: TestPutBlock (2.05s)
	tdp_test.go:187: read tcp 127.0.0.1:32850->127.0.0.1:40000: i/o timeout
FAIL
FAIL	github.com/coreos/torus/distributor/protocols/tdp	4.745s
```

### PR DESCRIPTION
**Pull request by ericchiang**
This PR turns on the race detector for all go tests and fixes some
code that currently triggers race errors.

closes #201

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -11,7 +11,7 @@ build: vendor
 	go build -o bin/torusblk -ldflags "-X $(REPOPATH).Version=$(VERSION)" ./cmd/torusblk
 
 test: bin/glide
-	go test $(shell ./bin/glide novendor)
```
Comment (xiang90) [inline Makefile]: --race?

**Comment by xiang90** [comment]
lgtm

File: Makefile (hunk 1 above)
Comment (ericchiang) [inline Makefile]: Go flags. I think it's just one dash. Plus this definitely runs the race detector

Comment (xiang90) [inline Makefile]: sure. but go default flag can also recognize -- for long flag. we prefer -- over - in our docs. so probably we should also use -- in scripts. but not a big deal.

Comment (ericchiang) [inline Makefile]: No problem. Updated.

**Comment by philips** [comment]
LGTM

**Comment by alpe** [comment]
To fail on a race condition `halt_on_error=1` should be set.
https://golang.org/doc/articles/race_detector.html

**Comment by ericchiang** [comment]
@alpe the race detector seems to already correctly fails tests without this set. 

```
$ echo $GORACE # not set
$ cat race_test.go 
package main

import (
	"sync"
	"testing"
)

func TestRace(t *testing.T) {
	var (
		a  int
		wg sync.WaitGroup
	)
	n := 100
	wg.Add(n)
	for i := 0; i < n; i++ {
		go func() { a += 1; wg.Done() }()
	}
	wg.Wait()
}
$ go test -v --race race_test.go
...
--- PASS: TestRace (0.01s)
PASS
Found 2 data race(s)
exit status 66
FAIL	command-line-arguments	1.012s
$ echo $?
1
```

**Comment by alpe** [comment]
@ericchiang without the `GORACE` setting the test will pass but the test run will fail. You are right. I got used to have the test fail as well but this are just my settings. Sorry, for the confusion.

**Comment by ericchiang** [comment]
@barakmich surprisingly this doesn't conflict with #226 but I've rebased for good measure. Ready whenever you are.

**Comment by barakmich** [comment]
LGTM, Merge on Green

**Pull request merged**